test/new-e2e/tests/agent-shared-components/forwarder/dns_failover_test.go

```diff
@@ -0,0 +1,277 @@
+
+	// configure agent to use the custom intake, set connection_reset_interval, use logs, and processes
+	agentOptions := []agentparams.Option{
+		agentparams.WithAgentConfig(getAgentConfig(intakeName, connectionResetInterval)),
+		agentparams.WithLogs(),
+		agentparams.WithIntegration("custom_logs.d", customLogsConfig),
+	}
+	v.UpdateEnv(multiFakeintakeStackDef(agentOptions...))
+
+	// check that fakeintake1 does receive metrics
+	v.T().Logf("checking that the agent contacts main intake at %s", fakeintake1IP)
+	require.EventuallyWithT(
+		v.T(),
+		assertIntakeIsUsed(v.Env().VM, v.Env().Fakeintake1, v.Env().Agent),
+		intakeMaxWaitTime,
+		intakeTick,
+	)
+
+	// check that fakeintake2 doesn't receive metrics
+	assertIntakeNotUsed(v.T(), v.Env().VM, v.Env().Fakeintake2, v.Env().Agent)
+
+	// perform local version of DNS failover
+	fakeintake2IP, err := hostIPFromURL(v.Env().Fakeintake2.URL())
+	v.NoError(err)
+	setHostEntry(v.T(), v.Env().VM, intakeName, fakeintake2IP)
+
+	// check that fakeintake2 receives metrics
+	v.T().Logf("checking that the agent contacts fallback intake at %s", fakeintake2IP)
+	require.EventuallyWithT(
+		v.T(),
+		assertIntakeIsUsed(v.Env().VM, v.Env().Fakeintake2, v.Env().Agent),
+		connectionResetInterval*time.Second+intakeMaxWaitTime,
+		intakeTick,
+	)
+}
+
+// assertIntakeIsUsed asserts the the given intakes receives metrics, logs, and flares
+func assertIntakeIsUsed(vm *client.VM, intake *client.Fakeintake, agent *client.Agent) func(*assert.CollectT) {
+	return func(t *assert.CollectT) {
+		// check metrics
+		metricNames, err := intake.GetMetricNames()
+		require.NoError(t, err)
+		assert.NotEmpty(t, metricNames)
+
+		// check logs
+		vm.Execute("echo 'totoro' >> /tmp/test.log")
+		logs, err := intake.FilterLogs("custom_logs")
+		require.NoError(t, err)
+		assert.NotEmpty(t, logs)
+
+		// check flares
+		agent.Flare(client.WithArgs([]string{"--email", "e2e@test.com", "--send"}))
+		_, err = intake.GetLatestFlare()
+		if err != nil {
+			require.ErrorIs(t, err, fi.ErrNoFlareAvailable)
+		}
+		assert.NoError(t, err)
+	}
+}
+
+// assertIntakeNotUsed asserts that the given intake doesn't receive metrics, logs, and flares
+func assertIntakeNotUsed(t *testing.T, vm *client.VM, intake *client.Fakeintake, agent *client.Agent) {
+	// check metrics
+	metricNames, err := intake.GetMetricNames()
+	require.NoError(t, err)
+	require.Empty(t, metricNames)
+
+	// check logs
+	vm.Execute("echo 'totoro' >> /tmp/test.log")
+	logs, err := intake.FilterLogs("custom_logs")
+	require.NoError(t, err)
+	require.Empty(t, logs)
+
+	// check flares
+	agent.Flare(client.WithArgs([]string{"--email", "e2e@test.com", "--send"}))
+	_, err = intake.GetLatestFlare()
+	require.ErrorIs(t, err, fi.ErrNoFlareAvailable)
+}
+
+// setHostEntry adds an entry in /etc/hosts for the given hostname and hostIP
+// if there is already an entry for that hostname, it is replaced
+func setHostEntry(t *testing.T, vm *client.VM, hostname string, hostIP string) {
```

Comment:
Strictly speaking this is an NSS validation, and not DNS. This is probably OK since the hostname lookup will typically use the NSS switch and perform a DNS lookup should no entry be available in any other source (like `/etc/hosts`). 

Though probably not necessary, you may want to update some comments to be more precise.

Reply:
That's a good remark, I'll update my comments !

Makes me think that when I started working on this test I checked manually that the `nsswitch.conf` file of my VM used `files` before `dns`, I should enforce that in the test

My take is that if we enforce that `nsswitch.conf` first uses `files` then it's fine not to test a real *DNS\* change, what do you think ?

Reply:
> I should enforce that in the test

Done https://github.com/DataDog/datadog-agent/pull/20167/commits/7a64976d43f1663074778e263d57f0d9497c9311